File: VocalesAndroid/app/src/main/java/app/badrs/vocales/MainActivity.java
package app.badrs.vocales;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.pm.PackageManager;
import android.graphics.Typeface;
import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.os.Bundle;
import android.text.Html;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetSocketAddress;

public class MainActivity extends AppCompatActivity {

    /* Views */
    private ImageView imageViewStreaming;
    private TextView textViewIsStreaming;
    private Button buttonStream;

    /* Configuration */
    private Thread streamingThread;
    private final int PORT = 55286;
    private boolean isStreaming = false;

    /* UDP Socket */
    private DatagramSocket udpServerSocket;

    /* AudioRecorder + Audio Configuration */
    private AudioRecord audioRecorder;
    private final int RECORDER_SOURCE = MediaRecorder.AudioSource.MIC;
    private final int SAMPLE_RATE = 44100;
    private final int CHANNEL = AudioFormat.CHANNEL_IN_MONO;
    private final int FORMAT = AudioFormat.ENCODING_PCM_16BIT;
    private final int BUFFER_SIZE = AudioRecord.getMinBufferSize(SAMPLE_RATE, CHANNEL, FORMAT);
    private final byte[] buffer = new byte[BUFFER_SIZE];

    @SuppressLint("SetTextI18n")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        /* Get imageViewStreaming */
        imageViewStreaming = findViewById(R.id.imageViewStreaming);

        /* Get textViewIsStreaming */
        textViewIsStreaming = findViewById(R.id.textViewIsStreaming);

        /* Get buttonStream and set onClickListener */
        buttonStream = findViewById(R.id.buttonStream);
        buttonStream.setOnClickListener(view -> {
            if (isStreaming) {
                imageViewStreaming.setVisibility(View.INVISIBLE);

                textViewIsStreaming.setText(R.string.text_is_streaming);

                buttonStream.setText(R.string.text_stream);
                buttonStream.setBackgroundColor(getResources().getColor(R.color.purple_500));

                // Stop the thread for socket streaming
                isStreaming = false;
                stopStreaming();

            } else {
                imageViewStreaming.setVisibility(View.VISIBLE);

                String streamingText = "Streaming to port <b>" + PORT + "</b>..";
                textViewIsStreaming.setText(Html.fromHtml(streamingText));
                textViewIsStreaming.setTypeface(textViewIsStreaming.getTypeface(), Typeface.BOLD);

                buttonStream.setText("Stop stream");
                buttonStream.setBackgroundColor(getResources().getColor(R.color.teal_700));

                // Start actual socket streaming
                isStreaming = true;
                verifyRecordAudioPermissionsThenStream();
            }
        });
    }

    private void startStreaming() {

        /* This disables the button access for 2 seconds */
        Utility.temporarilyDisableButtonAccess(this, buttonStream, 2000);

        /* Start UDP socket server */
        streamingThread = new Thread(() -> {
            try {
                /* We pass null in the UDP Socket constructor so that we can bind
                 IPv4 on the socket later */
                udpServerSocket = new DatagramSocket(null);

                /* Get the IPv4 of the device */
                InetSocketAddress address = Utility.getLocalAddress(MainActivity.this, PORT);

                udpServerSocket.bind(address);

                runOnUiThread(() ->
                    Utility.toast(this, udpServerSocket.getLocalSocketAddress().toString())
                );

                /* Initialize audioRecorder and start recording */
                audioRecorder = new AudioRecord(RECORDER_SOURCE, SAMPLE_RATE, CHANNEL, FORMAT,
                                                BUFFER_SIZE * 10);
                audioRecorder.startRecording();

                /* Send audio data while the thread is stopped -> stopStreaming() is called */
                while (streamingThread == Thread.currentThread()) {
                    /* Check if client is connected */
                    DatagramPacket client = new DatagramPacket(new byte[0], 0);
                    udpServerSocket.receive(client);

                    /* Read audioRecorder data and wrap it in a packet for UDP socket to send */
                    int recorderRead = audioRecorder.read(buffer, 0, buffer.length);
                    DatagramPacket packet = new DatagramPacket(
                            buffer,
                            recorderRead,
                            client.getAddress(),
                            client.getPort());
                    udpServerSocket.send(packet);
                }
            } catch (IOException error) {
                error.printStackTrace();
            }
        });
        streamingThread.start();
    }

    private void stopStreaming() {
        streamingThread = null;
         /* Stopping audio after setting thread to null  will avoid crashing app */
        udpServerSocket.close();

        Utility.toast(this, "Closed socket");
        Utility.temporarilyDisableButtonAccess(this, buttonStream, 2000);
    }

    private void verifyRecordAudioPermissionsThenStream() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO)
                != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{
                    Manifest.permission.RECORD_AUDIO
            }, 1);
        } else {
            startStreaming();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 1) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                startStreaming();
            } else {
                Utility.toast(this, "No permission to record audio");
            }
        }
    }
}

File: VocalesAndroid/app/src/main/java/app/badr/vocalesandroid/MainActivity.kt
package app.badr.vocalesandroid

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.net.wifi.WifiManager
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.WindowCompat
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.navigateUp
import androidx.navigation.ui.setupActionBarWithNavController
import app.badr.vocalesandroid.databinding.ActivityMainBinding
import com.google.android.material.snackbar.Snackbar
import java.math.BigInteger
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress

class MainActivity : AppCompatActivity() {

  private lateinit var appBarConfiguration: AppBarConfiguration
  private lateinit var binding: ActivityMainBinding

  private lateinit var textViewIpAddress: TextView
  private lateinit var textViewPacket: TextView

  private var requestCodeMicrophone = 100


  // Audio Recorder config
  private lateinit var audioRecorder: AudioRecord
  private var recorderSource: Int = MediaRecorder.AudioSource.MIC
  private val sampleRate = 44100
  private val channel: Int = AudioFormat.CHANNEL_IN_MONO
  private val format: Int = AudioFormat.ENCODING_PCM_16BIT
  private val bufferSize = AudioRecord.getMinBufferSize(sampleRate, channel, format)
  private val buffer = ByteArray(bufferSize)

  // Streaming thread
  private val localHostPort: Int = 1337
  private var isStreaming: Boolean = false
  private lateinit var streamingThread: Thread

  @RequiresApi(Build.VERSION_CODES.M)
  override fun onCreate(savedInstanceState: Bundle?) {
    WindowCompat.setDecorFitsSystemWindows(window, false)
    super.onCreate(savedInstanceState)

    binding = ActivityMainBinding.inflate(layoutInflater)
    setContentView(binding.root)

    setSupportActionBar(binding.toolbar)

    val navController = findNavController(R.id.nav_host_fragment_content_main)
    appBarConfiguration = AppBarConfiguration(navController.graph)
    setupActionBarWithNavController(navController, appBarConfiguration)

    // Views
    textViewIpAddress = findViewById(R.id.textViewIpAddress)
    textViewPacket = findViewById(R.id.textViewPacket)

    binding.fab.setOnClickListener { view ->
      if (isStreaming) {
        streamingThread.interrupt()
        isStreaming = false
        Toast.makeText(this, "Stopping stream", Toast.LENGTH_SHORT).show()

        return@setOnClickListener
      }

      when (ContextCompat.checkSelfPermission(
        this, Manifest.permission.RECORD_AUDIO
      )) {
        PackageManager.PERMISSION_GRANTED -> {
          enableMicrophone()
          sendStream()
        }
        PackageManager.PERMISSION_DENIED -> {
          this.requestPermissions(
            arrayOf(Manifest.permission.RECORD_AUDIO),
            requestCodeMicrophone
          )
        }
      }
      Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
        .setAnchorView(R.id.fab)
        .setAction("Action", null).show()
    }
  }

  @SuppressLint("MissingPermission")
  private fun enableMicrophone() {
    audioRecorder = AudioRecord(recorderSource, sampleRate, channel, format, bufferSize)
    audioRecorder.startRecording()
  }

  private fun sendStream() {
    // Make datagram socket and bind address as null initially
    val datagramSocket = DatagramSocket(null)

    // Bind IPv4 address of the device to the socket
    val wifiManager = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
    val localHostAddress = wifiManager.connectionInfo.ipAddress
    val localHostAddressString = InetAddress.getByAddress(
      BigInteger.valueOf(localHostAddress.toLong()).toByteArray()
    ).hostAddress

    // Bind socket to the local host and set view
    datagramSocket.bind(
      InetSocketAddress(localHostAddressString, localHostPort)
    )

    textViewIpAddress.append(localHostAddressString)

    streamingThread = Thread {
      while (true) {
        // Get microphone stream bytes
        val recorderBytes = audioRecorder.read(buffer, 0, buffer.size)

        // Make datagram packet and send
        val packet = DatagramPacket(buffer, recorderBytes)
        datagramSocket.send(packet)

        textViewPacket.text = "${buffer.toString().subSequence(0, 5)}"
      }
    }

    streamingThread.start()
    isStreaming = true
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    // Inflate the menu; this adds items to the action bar if it is present.
    menuInflater.inflate(R.menu.menu_main, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    // Handle action bar item clicks here. The action bar will
    // automatically handle clicks on the Home/Up button, so long
    // as you specify a parent activity in AndroidManifest.xml.
    return when(item.itemId) {
      R.id.action_settings -> true
      else -> super.onOptionsItemSelected(item)
    }
  }

  override fun onSupportNavigateUp(): Boolean {
    val navController = findNavController(R.id.nav_host_fragment_content_main)
    return navController.navigateUp(appBarConfiguration)
            || super.onSupportNavigateUp()
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ) {
    when (requestCode) {
      requestCodeMicrophone -> {
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
          enableMicrophone()
        } else {
          Toast.makeText(
            this,
            "Enable your microphone to use this feature",
            Toast.LENGTH_LONG).show()
        }
      } else -> {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
      }
    }
  }
}